<a href="https://colab.research.google.com/github/stoicelephant/DIDA325/blob/main/Copy_of_325_Fall_Midterm_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h3>I am first going to import all of the neccessary libraries that I will need to work with dataframes (pandas) and for data visualizations (matplotlib and seaborn) </h3>

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

<h3>I am now going to load in the dataset that I will be working on </h3)

In [ ]:
voter_reg = pd.read_csv('https://raw.githubusercontent.com/fivethirtyeight/data/refs/heads/master/voter-registration/new-voter-registrations.csv')

##Introduction (10 points)
You've been hired to work with a government organization interested in formation about voter registration in the United States. They have asked you to come up with two research questions--one about national trends and one about a state trend of your choice.

Clearly state:
* the purpose of your project
* what the goal is
* why it is important for someone to read this notebook?

In additon, state your two research questions

* Question 1:
* Question 2:



<h3>Task: Formulate a Research question regarding voter registration trends on the (1) State and (2) National Level</h3>
<h3>- Question 1: </h3>
<h3>- Question 2: </h3>

<h3>Purpose: The purpose of my project is...</h3>
<h3>Goal: The goal of my project is... </h3>
<h3>This Notebook provides in depth analysis as well as visualiztions surrounding state and national voter registration trends</h3>

<h3>Below is a preview of the data</h3>

In [ ]:
voter_reg.head()

Jurisdiction  Year Month  New registered voters
0      Arizona  2016   Jan                  25852
1      Arizona  2016   Feb                  51155
2      Arizona  2016   Mar                  48614
3      Arizona  2016   Apr                  30668
4      Arizona  2020   Jan                  33229

<h3> As we can see the data is comprised of 5 different columns. Namely; <br>
1) Juridiction - the state of interest <br>
2) Year - the year of interest <br>
3) Month - the month of interest <br>
4) New Registered Voters - the number of newly registered voters located in the juisdicitional state of interest during the respective year and month of interest. </h3>

<h3>I am now going to check the table to see if theres any empty/null values that I will have to deal with. As can be seen in the code cell below, there is no N/A cells in our dataframe. </h3>

In [ ]:
voter_reg.isna().sum()

Jurisdiction             0
Year                     0
Month                    0
New registered voters    0
dtype: int64

<h3>To get a better idea of the data we are working with, I am going to check the data types of each of the columns. As can be seen in the code cell below, <br> 1) Jurisdiction - an object / string datatype <br> 2) Year - a 64 bit integer datatype <br> 3) Month - an object / string datatype <br> 4) New registered voters - a 64 bit integer datatype  </h3>

In [ ]:
voter_reg.dtypes

Jurisdiction             object
Year                      int64
Month                    object
New registered voters     int64
dtype: object

<h3>To make my life a little easier when referencing the columns of the table, I am going to rename the 'jurisdiction' column to 'NAME' and the 'New registered voters' to 'registered_voters' </h3>

In [ ]:
voter_reg = voter_reg.rename(columns={'Jurisdiction':'NAME','New registered voters':'registered_voters'})
voter_reg

NAME  Year Month  registered_voters
0     Arizona  2016   Jan              25852
1     Arizona  2016   Feb              51155
2     Arizona  2016   Mar              48614
3     Arizona  2016   Apr              30668
4     Arizona  2020   Jan              33229
..        ...   ...   ...                ...
101  Virginia  2020   Jan              25934
102  Virginia  2020   Feb              29507
103  Virginia  2020   Mar              31492
104  Virginia  2020   Apr               5467
105  Virginia  2020   May               8239

[106 rows x 4 columns]

Explain any potential limitations to your data (2 points)

Limitation of My Data: <br>
1) There are only 106 rows / lines of data in my DataFrame. This isn't a ton of data, making it difficult to find meaninful trends.
2) As can be seen in the line of code below we must be aware that we are working specifically with data from 11 distinct states and 1 federal district ( The District of Columbia) - Additionally back to the first point there is only a few pieces of data for each of these locations. Also, most of these states are in the southern half of the US.
3) We only have data on the electoral years of 2016 and 2020, during the electoral months of January (01) -May (05)

In [ ]:
voter_reg.NAME.unique()

array(['Arizona', 'California', 'Colorado', 'Delaware',
       'District of Columbia', 'Florida', 'Georgia', 'Illinois',
       'Maryland', 'North Carolina', 'Texas', 'Virginia'], dtype=object)

In [ ]:
voter_reg.Year.unique()

array([2016, 2020])

In [ ]:
voter_reg.Month.unique()

array(['Jan', 'Feb', 'Mar', 'Apr', 'May'], dtype=object)

<h3>I am now going to perform some exploratory data analysis to get a better grasp of what is going on, and try and see if I can find some preliminary trends I'd like to analyze further. </h3>
<h4>As can be seen in the data below i get an idea on the distribution of the voter registration counts, specifically finding the max as well as a measure of central tendency, specifically <br>

I compute $$\sum_{n\in N}{\text{registered_voters}_n}/N$$ where n is each of the rows in the data. I obtained that on average, approximately 48,200 people were newely registered voters per state per voting month.
 </h4>

In [ ]:
voter_reg.loc[voter_reg.registered_voters == voter_reg.registered_voters.max()]

NAME  Year Month  registered_voters
13  California  2020   Feb             238281

In [ ]:
print(float(voter_reg.registered_voters.mean()))

48223.46226415094


Create a dataframe called "high_voters" showing only the times where voter registration was higher than your above average. How many times did this happen? (3 points)


How many times did each state go above the national average for new voter registration? (2 points)



Which three states had the highest average voter registration overall? Show only the top three results. (5 points)

##Data Visualization (30 points, as marked)

Create a plot showing the voter registratation by state.
* A dataframe/a way to show the total voters registered per state. If you create a new dataframe, you may need to include `reset_index()`.(3 points)
* Hide the legend by passing legend=false through the plot code (1 point)
* Clearly legible labels and titles (2 points)
* A color palette that is not the default palette and that makes sense for your data. (1 point)
* Comment on the trends you see (2 points)
(9 points total)

Produce a plot comparing voter registation in 2016 and 2020. (2 pts)
* Color the graph based on the month.(1 pt)
* Change the default color palette used. (1 pt)
* Comment on any trends you see. (2 pts)
* Add the appropriate labels and title (1)
* Comment on any trends you see (2 points)

(9 pts total)

Create a data visualization that relates to either your state level research question or your national level reserach question. If one of your questions was answered by the above graphs, you may not use that question. (12 points)

##Notebook Conclusion  (15 points)

Write a conclusion section that includes
* Insights: the insights/outcomes of your notebook (5 points)
* Suggestions: Any suggestions or ideas you could offer your client (5 point)
* Possible next steps: At least one step that you would take if continuing to work with this project (5 points)

This should only be reflective of *your data* as presented here.


Each section should be at least 3-4 complete sentences.


##Stakeholder Analysis(14 points as noted)
*Loosely adapted from the [Cambridge Analytica/Facebook reveal](https://www.google.com/search?q=cambridge+analytica+facebook+&sca_esv=d43255849d412494&rlz=1C1ONGR_enUS1069US1069&sxsrf=AE3TifP9Zpi1D8vk3Z__wLxSQ-IMSAubZg%3A1758661179884&ei=OwrTaMHdNYmx5NoPkdTiqQY&ved=0ahUKEwiB56C34--PAxWJGFkFHRGqOGUQ4dUDCBA&uact=5&oq=cambridge+analytica+facebook+&gs_lp=Egxnd3Mtd2l6LXNlcnAiHWNhbWJyaWRnZSBhbmFseXRpY2EgZmFjZWJvb2sgMgoQIxiABBgnGIoFMgUQABiABDIIEAAYgAQYxwMyCBAAGIAEGMcDMggQABiABBjHAzIFEAAYgAQyCBAAGIAEGMcDMgYQABgWGB4yBhAAGBYYHjIGEAAYFhgeSJoJUI4DWI4DcAF4AZABAJgBa6ABa6oBAzAuMbgBA8gBAPgBAZgCAqAChAHCAgoQABiwAxjWBBhHmAMAiAYBkAYIkgcDMS4xoAelCLIHAzAuMbgHc8IHBTItMS4xyAcV&sclient=gws-wiz-serp) in 2018. While it is not necessary to look this up, it's always good to know*.

All answers should be in full sentences and in a text box, not code.

You work for a social media company which is currently allowing third party apps access to data through the use of an API; this means another company can create an app on your platform and acces the data of users who consent.

One such company is a lobbying group. They create a third party app called "Get Out and Vote!" that asks users about their voter registration and other habits. While users of the app must give consent, the app also allows the lobbying group to collect data on the social media habits of the friends of those who consented to the app. While only only a few hundred thousand partake in the original app, the company is still able to collect from millions of users.

They use this data to create personality profiles of unregistered voters. That data is then sold to specific political campaigns who use the information to create political ads that encourage people in populated US cities *not to vote* while creating ads that tell people in swing states *to vote*.

While it is not necessary to do research on swing states, here is some information about the demographics of that if you are [curious](https://www.npr.org/2024/09/06/nx-s1-5101765/swing-states-demographics-electorate).


Identify a key ethical issue here, from the standpoint of the social media company (3 points).

What are at least three of our ethical standards that apply to this situation, and why are they relevant (5 points).

Your company wants to continue work with the lobbying firm and does not want to lose them as a client. Looking at this from a *utilitarian ethics* standpoint, propose an alternative action that allows you to work with them while still addressing the ethical issue you mentioned above (6 points).



##Github (3 points)
Post this to your Github and include the accessible link either here or on brightspace.

###You can also find my project in my [github](https://github.com/stoicelephant/DIDA325/blob/main/325_Fall_Midterm_Project.ipynb)

# Academic Integrity Statement

By writing my name in the cell below, I certify that:

1. I did not use resources other than:
   * the Python notebooks provided by the instructor,
   * links provided in this notebook,
   * the assigned readings, and
   * my own personal notes
2. This means that I did not:
   * look up anything on Google, or Stack Overflow, Chatgpt, &c.,
   * discuss the content of the exam with *anyone* other than the instructors or TAs, or
   * do anything that technically doen't break these rules but is against their spirit.

In [1]:
James Giuffre

SyntaxError: invalid syntax (ipython-input-3453925861.py, line 1)